src/Workspaces/Remote/Core/RemoteCallback.cs

```diff
@@ -87,23 +88,106 @@ public async ValueTask<TResult> InvokeAsync<TResult>(Func<T, CancellationToken,
-        /// The API streams results back to the caller.
+        /// Invokes API on the callback object hosted in the original process (usually devenv) associated with the
+        /// currently executing brokered service hosted in ServiceHub process. The API streams results back to the
+        /// caller.
         /// </summary>
-        /// <inheritdoc cref="BrokeredServiceConnection{TService}.InvokeStreamingServiceAsync"/>
+        /// <param name="invocation">A callback to asynchronously write data. The callback should always <see
+        /// cref="PipeWriter.Complete"/> the <see cref="PipeWriter"/>.  If it does not then reading will hang</param>
+        /// <param name="reader">A callback to asynchronously read data. The callback should not complete the <see
+        /// cref="PipeReader"/>, but no harm will happen if it does.</param>
+        /// <param name="cancellationToken">A cancellation token the operation will observe.</param>
         public async ValueTask<TResult> InvokeAsync<TResult>(
             Func<T, PipeWriter, CancellationToken, ValueTask> invocation,
             Func<PipeReader, CancellationToken, ValueTask<TResult>> reader,
             CancellationToken cancellationToken)
         {
             try
             {
-                return await BrokeredServiceConnection<T>.InvokeStreamingServiceAsync(_callback, invocation, reader, cancellationToken).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+                var pipe = new Pipe();
+
+                // Kick off the work to do the writing to the pipe asynchronously.  It will start hot and will be able
+                // to do work as the reading side attempts to pull in the data it is writing.
+
+                var writeTask = WriteAsync(_callback, pipe.Writer);
+                var readTask = ReadAsync(pipe.Reader);
+
+                // Note: waiting on the write-task is not strictly necessary.  The read-task cannot complete unless it
+                // the write-task completes (or it faults for some reason).  However, it's nice and clean to just not
+                // use fire-and-forget here and avoids us having to consider things like async-tracking-tokens for
+                // testing purposes.
+                await Task.WhenAll(writeTask, readTask).ConfigureAwait(false);
+                return await readTask.ConfigureAwait(false);
             }
             catch (Exception exception) when (ReportUnexpectedException(exception, cancellationToken))
             {
                 throw new OperationCanceledIgnoringCallerTokenException(exception);
             }
+
+            async Task WriteAsync(T service, PipeWriter pipeWriter)
+            {
+                try
+                {
+                    // Intentionally yield this thread so that the caller can proceed concurrently and start reading.
+                    await Task.Yield();
```

Comment:
```suggestion
                    // Intentionally yield this thread so that the caller can proceed concurrently and start reading.
                    // This is not strictly necessary (as we know the writer will always call FlushAsync()), but it is nice
                    // as it allows both to proceed concurrently on the initial writing/reading.
                    await Task.Yield();
```